daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerInteraction.scala

```diff
@@ -461,44 +438,19 @@ class IdeClient(val compiledPackages: CompiledPackages) extends ScriptLedgerClie
       }
   }
 
-  // Translate from a ledger command to an Update expression
-  // corresponding to the same command.
-  private def translateCommand(cmd: ScriptLedgerClient.Command): speedy.Command = {
-    // Ledger commands like create or exercise look pretty complicated in
-    // SExpr. Therefore we express them in the high-level AST and compile them
-    // to a function that we apply to the arguments.
-    cmd match {
-      case ScriptLedgerClient.CreateCommand(tplId, arg) =>
-        speedy.Command.Create(tplId, arg)
-      case ScriptLedgerClient.ExerciseCommand(tplId, cid, choice, arg) =>
-        speedy.Command.Exercise(tplId, SContractId(cid), choice, arg)
-      case ScriptLedgerClient.CreateAndExerciseCommand(tplId, tpl, choice, arg) =>
-        speedy.Command.CreateAndExercise(tplId, tpl, choice, arg)
-      case ScriptLedgerClient.ExerciseByKeyCommand(tplId, key, choice, arg) =>
-        speedy.Command.ExerciseByKey(tplId, key, choice, arg)
-    }
-  }
-
-  // Translate a list of commands submitted by the given party
-  // into an expression corresponding to a scenario commit of the same
-  // commands of type `Scenario ()`.
-  private def translateCommands(commands: List[ScriptLedgerClient.Command]): SExpr = {
-    val cmds: ImmArray[speedy.Command] = ImmArray(commands.map(translateCommand(_)))
-    compiledPackages.compiler.unsafeCompile(cmds)
-  }
-
   // unsafe version of submit that does not clear the commit.
   private def unsafeSubmit(
       party: SParty,
-      commands: List[ScriptLedgerClient.Command],
+      commands: List[command.Command],
       optLocation: Option[Location])(implicit ec: ExecutionContext)
     : Future[Either[StatusRuntimeException, Seq[ScriptLedgerClient.CommandResult]]] = Future {
     // Clear state at the beginning like in SBSBeginCommit for scenarios.
     machine.commitLocation = optLocation
     machine.returnValue = null
     machine.localContracts = Map.empty
     machine.globalDiscriminators = Set.empty
-    val translated = translateCommands(commands)
+    val speedyCommands = preprocessor.unsafePreprocessCommands(commands.to[ImmArray])._1
```

Comment:
This seems strictly slower than what we did before: We get an `SValue` for all commands from speedy. Then you convert it to a `Value` only to then convert it back to `SValue`s here. While you can argue that this is not a noticeable performance loss, I’m at least somewhat cautious in doing this without having measured it.